contents/ai/ml.md

```diff
@@ -0,0 +1,149 @@
+4 mistakes in fraud detection is a much bigger issue than 4 mistakes in hotdog classification. Assuming a validation set of 1000, 4 mistakes equates to 0.2% difference in accuracy. If working on a business problem like hotdog classification, comparing models with a small performance difference like in this example will require a bigger validation set.   
+
+- Statistical Stability  
+Each class should have at least 22 datapoints, so that the validation set follows an approximate normal distribution. This allows easier statistical analysis for the next driving factor.
+
+- Standard Error   
+The standard error, calculated using statsistical analysis, should be smaller than the difference in performance of models. This helps us ensure that the difference in performance is reliable.
+
+See Lesson 7 of FastAI's ML MOOC for a more in-depth explanation (MOOC is at unofficial release at time of writing)
+
+#### Training Set
+The remaining data forms the training set. Data in the training set is used for model training. 
+
+Practitioners may also choose to create a sample set out of the training set, with a much smaller amount of data. They then train their models on the sample set rather than training set during this prototyping phase. This allows for quicker testing and refinement iterations. Training on the entire training set occurs only after they discover a set of parameters they feel confident with. 
+
+A good article on test and validation sets can be found [here](http://www.fast.ai/2017/11/13/validation-sets/).
+
+### Model Training, Evaluation & Data Analysis
+The model is evaluated on the validation set after it is trained.
+Practitioners also perform data analysis on the results to gain further insights on the data.
+These insights guide the practitioner in tweaking the data and model for better performance.
+
+An example would be noticing that the model consistently fails to perform for low resolution images. All images have been resized to a standard size at the preprocessing phase. A potential fix could be to downscale all images so no image is at a low resolution, training the model on them, and then continuing to train on the images at their initial preprocessed sizes. 
+
+Looking for highly correlated fields and investigating their impact is also a common procedure when analyzing structured data. The model can then be retrained with correlated fields removed if they do not contribute much to the model's predictions. Removal of such fields usually improves model performance. Tools for statistical analysis, such as confusion matrices and Spearman's rho are often used here. 
+
+![confusion matrix](http://scikit-learn.org/stable/_images/sphx_glr_plot_confusion_matrix_001.png)  
+(image from scikit-learn's documentation)
+
+The confusion matrix above shows that the model has predicted 6 instances of 'versicolor' to be 'virginica'. These misclassifications can be investiagated to obtain insights on both model performance and data.
```

Comment:
"investiagated" is spelled wrongly.

Reply:
Are there ways/tools to easily investigate the misclassifications?

Reply:
The most basic way to investigate would be to look at the exact examples that failed. Not sure if that's worth mentioning since investigating should already mean that.
The more complex ones are model specific.